web3/providers/auto.py

```diff
@@ -12,20 +26,26 @@
     IPCProvider,
     WebsocketProvider,
 )
+from web3.types import (
+    RPCEndpoint,
+    RPCResponse,
+)
 
 HTTP_SCHEMES = {'http', 'https'}
 WS_SCHEMES = {'ws', 'wss'}
 
 
-def load_provider_from_environment():
-    uri_string = os.environ.get('WEB3_PROVIDER_URI', '')
+def load_provider_from_environment() -> BaseProvider:
```

Comment:
Not necessarily for this PR but it might be good to create a `web3.abc` module with a `ProviderAPI` at some point soon.  This will be especially valuable for other libraries that need to do type hinting.